[SecuritySolution] Move related alerts by process ancestry behind platinum subscription

janmonschke opened the pull request:
## Summary

We're making `alerts by process ancestry` a platinum+ feature.
Only clusters with an active trial sub or a platinum+ sub will be able to access that feature. All other users will see an upsell in the insights section of the alert flyout.

| Platinum user | Basic user | 
|--|--|
| <img width="670" alt="Screenshot 2022-08-09 at 17 09 15" src="https://user-images.githubusercontent.com/68591/183688810-d2b515c6-98bf-451c-8089-daa50e5f9f3c.png"> | <img width="992" alt="Screenshot 2022-08-09 at 17 09 48" src="https://user-images.githubusercontent.com/68591/183688817-1c1a943c-c8be-4ff9-9b88-9ac01eab47cb.png"> |

### Test instructions
In order to test this feature with a platinum subscription, go to `Management` -> `Stack Management` -> `License Management` and start a `trial` subscription from there.
You can also revert back to a `basic` license there.

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

<!--ONMERGE {"backportTargets":["8.4"]} ONMERGE-->

janmonschke committed:
feat: show an upsell for users without platinum license

janmonschke committed:
test: add license-related tests

janmonschke committed:
feat: re-add the feature flag

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.tsx:
```diff
@@ -4,6 +4,7 @@
  * 2.0; you may not use this file except in compliance with the Elastic License
  * 2.0.
  */
+/* eslint-disable complexity */
```
nit: if this `eslint-disable complexity` is removed, the entire file is highlighted, but temporarily changing just one randomly-chosen block of conditionals:

```ts
    const canShowAtLeastOneInsight =
      hasCasesReadPermissions ||
      hasProcessEntityInfo ||
      hasSourceEventInfo ||
      hasProcessSessionInfo;
```

to

```ts
    const canShowAtLeastOneInsight = true;
```

was enough to remove the highlighting for the entire file. Consider extracting this block of code into a helper function in a separate file and removing the `eslint-disable complexity`.

janmonschke replied:
Ah yes, nice one. I added a type guard here https://github.com/elastic/kibana/pull/138419/commits/24131f7be52465b2863577638e32d14dfa21443b 🎉

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.tsx:
```diff
@@ -45,14 +49,14 @@ export const Insights = React.memo<Props>(
     );
-    const hasProcessEntityInfo =
-      isRelatedAlertsByProcessAncestryEnabled && processEntityField && processEntityField.values;
+    const hasProcessEntityInfo = processEntityField && processEntityField.values;
```
Would you be willing to confirm that we still want to remove the `isRelatedAlertsByProcessAncestryEnabled` check?

andrew-goldstein replied:
nvm, it just migrated south for the winter to this check:

```ts
const hasProcessSessionInfo =
      isRelatedAlertsByProcessAncestryEnabled && processSessionField && processSessionField.values;
```

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.tsx:
```diff
@@ -45,14 +49,14 @@ export const Insights = React.memo<Props>(
       data
     );
-    const hasProcessSessionInfo = processSessionField && processSessionField.values;
+    const hasProcessSessionInfo =
```
if the length of `processSessionField.values` in the following check is zero:

```ts
    const hasProcessSessionInfo =
      isRelatedAlertsByProcessAncestryEnabled && processSessionField && processSessionField.values;
```

should `hasProcessSessionInfo` still be true?

(The same question may also apply to the other places in this file where `.values` is tested for truthiness.)

janmonschke replied:
Similar to your comment above, moved this to a type guard. Nice catch :) https://github.com/elastic/kibana/pull/138419/commits/24131f7be52465b2863577638e32d14dfa21443b

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_by_session.tsx:
```diff
@@ -99,6 +100,7 @@ export const RelatedAlertsBySession = React.memo<Props>(
         state={state}
         text={getTextFromState(state, count)}
         renderContent={renderContent}
+        extraAction={<EuiBetaBadge size="s" label={BETA} color="subdued" />}
```
Thanks for adding the `Beta` badge.

For consistency while it's behind a feature flag, should `Related alerts by process ancestry` also have a `Beta` badge?

![related_alerts_by_process_ancestry](https://user-images.githubusercontent.com/4459398/183732804-cf60d08c-9059-45cc-8a34-7d9342e19a80.png)

janmonschke replied:
I agree, makes sense. Added here: https://github.com/elastic/kibana/pull/138419/commits/81ca4d08c0cacefa2d578bb8934b9c0c6ee4b175

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_upsell.tsx:
```diff
@@ -0,0 +1,44 @@
+
+export const RelatedAlertsUpsell = React.memo(() => {
+  return (
+    <UpsellContainer href="https://www.elastic.co/pricing/" target="_blank">
+      <EuiFlexGroup alignItems="center" gutterSize="none">
+        <EuiFlexItem grow={false}>
+          <StyledIcon size="m" type="lock" />
+        </EuiFlexItem>
+        <EuiFlexItem>
+          <EuiText size="s">{ALERT_UPSELL}</EuiText>
+        </EuiFlexItem>
+      </EuiFlexGroup>
+    </UpsellContainer>
```
Consider the following suggestion that:

- Replaces the `href` and `target` props to `UpsellContainer` with an `EuiLink`
- `EuiLink` is given `rel="noopener noreferrer"` as a standard practice for external links to both:
  - prevent [tabnabbing](https://cheatsheetseries.owasp.org/cheatsheets/HTML5_Security_Cheat_Sheet.html#tabnabbing) (unlikely given the source we're linking to)
  - prevent information leakage via the `Referer` HTTP request header
- `EuiLink` uses the `external={true}` prop, which renders an external link indicator icon

```suggestion
    <UpsellContainer>
      <EuiFlexGroup alignItems="center" gutterSize="none">
        <EuiFlexItem grow={false}>
          <StyledIcon size="m" type="lock" />
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiText size="s">
            <EuiLink
              color="subdued"
              external={true}
              href="https://www.elastic.co/pricing/"
              rel="noopener noreferrer"
              target="_blank"
            >
              {ALERT_UPSELL}
            </EuiLink>
          </EuiText>
        </EuiFlexItem>
      </EuiFlexGroup>
    </UpsellContainer>
```

If the suggestion above is implemented:

- The following styles in `UpsellContianer` are no longer required, because they are built into `EuiLink`:

```
  &:focus, &:hover {
    text-decoration: underline;
  }
```

- An external link icon will appear next to the link, per the before / after screenshots below:

**Before**

![with_styled_panel](https://user-images.githubusercontent.com/4459398/183761413-2e838884-6434-4dfb-86d6-af72d1a95b9a.png)

**After**

![with_eui_link](https://user-images.githubusercontent.com/4459398/183761460-c391d2d0-601c-4d6e-8ea9-652f75c739ce.png)

janmonschke replied:
Oh nice, I was not thinking about using an `EuiLink`, good to know that it has all the default styling and the correct referrer handling! https://github.com/elastic/kibana/pull/138419/commits/681b463a5b926b49f067a0d1436a1570233d629a

andrew-goldstein reviewed (approved):
Thanks for this update @janmonschke!

Desk tested locally with the following configurations:

### Trial license (platinum) with FF enabled:

![trial_license_ff_enabled](https://user-images.githubusercontent.com/4459398/183765411-b9eab5e9-1376-4f9b-a105-d20a5a478538.png)

### Trial license (platinum) with FF disabled:

![trial_license_ff_disabled](https://user-images.githubusercontent.com/4459398/183765540-766016cf-dd72-4feb-9da0-6b8fa0560d8d.png)

### Basic license with FF enabled:

![basic_license_ff_enabled](https://user-images.githubusercontent.com/4459398/183765626-5bf3f40c-d1dd-4292-b47a-d13a2e3a34e4.png)

### Basic license with FF disabled:

![basic_license_ff_disabled](https://user-images.githubusercontent.com/4459398/183765686-d6536292-6694-4975-b476-3e3b82cf797b.png)

LGTM 🚀

janmonschke committed:
nit: use EuiLink

janmonschke committed:
fix: check for length as well

janmonschke committed:
chore: move data check to helper

This reduced the overall complexity of the module, which made the eslint complexity exception obsolete

janmonschke committed:
tests:

janmonschke committed:
tests: fix tests

janmonschke committed:
tests: fix integration tests

janmonschke commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.tsx:
```diff
@@ -73,7 +71,6 @@ export const Insights = React.memo<Props>(
     const canShowAncestryInsight =
-      isRelatedAlertsByProcessAncestryEnabled &&
       processEntityField &&
       processEntityField.values &&
```
That one was necessary because TypeScript needed extra proof that the values exist, haha. I added a type guard in https://github.com/elastic/kibana/pull/138419/commits/24131f7be52465b2863577638e32d14dfa21443b so this code is not necessary anymore :)

janmonschke commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/translations.ts:
```diff
@@ -135,3 +135,10 @@ export const SIMPLE_ALERT_TABLE_LIMITED = i18n.translate(
+export const ALERT_UPSELL = i18n.translate(
+  'xpack.securitySolution.alertDetails.overview.insights.alertUpsellTitle',
+  {
+    defaultMessage: 'Get more insights with a subscription',
```
The translations are periodically added by another team. They're all here: https://github.com/elastic/kibana/tree/main/x-pack/plugins/translations/translations

logeekal replied:
Okay. Thank you.

banderror reviewed (approved):
Checked the diff and didn't find anything relevant to @elastic/security-detections-response-rules. Will revisit CODEOWNERS and try to unassign our area from irrelevant folders to not block folks.

andrew-goldstein commented:
## Files by Code Owner

### elastic/security-onboarding-and-lifecycle-mgt

* x-pack/plugins/security_solution/server/endpoint/routes/resolver.ts
* x-pack/plugins/security_solution/server/endpoint/routes/resolver/tree/handler.ts
* x-pack/plugins/security_solution/server/endpoint/services/feature_usage/service.ts

### elastic/security-solution

* x-pack/plugins/security_solution/public/common/components/event_details/insights/helpers.ts
* x-pack/plugins/security_solution/public/common/components/event_details/insights/insight_accordion.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.test.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_by_process_ancestry.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_by_session.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_upsell.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/translations.ts
* x-pack/plugins/security_solution/server/endpoint/routes/resolver.ts
* x-pack/plugins/security_solution/server/endpoint/routes/resolver/tree/handler.ts
* x-pack/plugins/security_solution/server/endpoint/services/feature_usage/service.ts

### elastic/security-threat-hunting-investigations

* x-pack/plugins/security_solution/public/common/components/event_details/insights/helpers.ts
* x-pack/plugins/security_solution/public/common/components/event_details/insights/insight_accordion.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.test.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_by_process_ancestry.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_by_session.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_upsell.tsx
* x-pack/plugins/security_solution/public/common/components/event_details/insights/translations.ts

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.4|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/138605">](https://github.com/elastic/kibana/pull/138605)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.0 -->
